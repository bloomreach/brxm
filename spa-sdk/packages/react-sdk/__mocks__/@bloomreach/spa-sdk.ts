/*
 * Copyright 2019-2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Component, Page } from '@bloomreach/spa-sdk';

const component = new class implements Component {
  getId = jest.fn();
  getMeta = jest.fn();
  getModels = jest.fn();
  getUrl = jest.fn();
  getName = jest.fn();
  getParameters = jest.fn();
  getChildren = jest.fn();
  getComponent = jest.fn();
  getComponentById = jest.fn();
};
const page = new class implements Page {
  getButton = jest.fn();
  getChannelParameters = jest.fn();
  getComponent = jest.fn(() => component);
  getContent = jest.fn();
  getDocument = jest.fn();
  getMeta = jest.fn();
  getTitle = jest.fn();
  getUrl = jest.fn();
  getVersion = jest.fn();
  getVisitor = jest.fn();
  getVisit = jest.fn();
  isPreview = jest.fn();
  rewriteLinks = jest.fn();
  sync = jest.fn();
  toJSON = jest.fn();
};

module.exports = {
  ...jest.genMockFromModule('@bloomreach/spa-sdk'),
  initialize: jest.fn(() => page),
};
