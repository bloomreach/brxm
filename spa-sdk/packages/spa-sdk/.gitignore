/coverage
/docs
/lib
/node_modules
