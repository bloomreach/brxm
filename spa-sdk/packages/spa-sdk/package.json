{
  "name": "@bloomreach/spa-sdk",
  "version": "14.3.1",
  "description": "Bloomreach SPA SDK",
  "keywords": [
    "bloomreach",
    "spa",
    "sdk"
  ],
  "author": "Bloomreach B.V.",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "https://code.onehippo.org/cms-community/bloomreach-spa-sdk",
    "directory": "packages/spa-sdk"
  },
  "bugs": {
    "url": "https://issues.onehippo.com/projects/CMS/issues"
  },
  "main": "dist/index.js",
  "browser": "dist/index.js",
  "module": "dist/index.mjs",
  "es2015": "dist/index.es6.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist/"
  ],
  "scripts": {
    "build": "rollup -c rollup.config.js",
    "docs": "typedoc",
    "lint": "tslint --project .",
    "prepare": "$npm_execpath run build",
    "test": "jest --coverage"
  },
  "jest": {
    "preset": "ts-jest",
    "reporters": [
      "default",
      "jest-junit"
    ],
    "setupFiles": [
      "reflect-metadata"
    ],
    "testEnvironment": "jsdom"
  },
  "jest-junit": {
    "outputDirectory": "coverage",
    "outputName": "TEST-spa-sdk.xml"
  },
  "browserslist": [
    "last 1 chrome version",
    "last 1 firefox version",
    "last 1 safari version",
    "last 1 edge version"
  ],
  "devDependencies": {
    "@babel/core": "^7.9",
    "@babel/plugin-proposal-class-properties": "^7.8",
    "@babel/plugin-proposal-object-rest-spread": "^7.9",
    "@babel/plugin-transform-object-assign": "^7.8",
    "@babel/polyfill": "^7.8",
    "@babel/preset-env": "^7.9",
    "@types/jest": "^25.2",
    "@types/xmldom": "^0.1",
    "babel-plugin-transform-async-to-promises": "^0.8",
    "jest": "^26.0",
    "jest-junit": "^11.1",
    "js-beautify": "^1.11",
    "rollup": "^2.10",
    "rollup-plugin-babel": "^4.4",
    "rollup-plugin-dts": "^1.4",
    "rollup-plugin-terser": "^5.3",
    "rollup-plugin-typescript2": "^0.27",
    "ts-jest": "^26.0",
    "tslint": "^6.1",
    "tslint-config-airbnb": "^5.11",
    "typedoc": "^0.17",
    "typedoc-plugin-external-module-map": "^1.2",
    "typescript": "^3.8"
  },
  "dependencies": {
    "emittery": "^0.7",
    "inversify": "^5.0",
    "reflect-metadata": "^0.1",
    "xmldom": "^0.3"
  }
}
