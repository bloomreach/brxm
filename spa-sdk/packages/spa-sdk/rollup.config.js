/*
 * Copyright 2019-2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import babel from 'rollup-plugin-babel';
import dts from 'rollup-plugin-dts';
import { terser } from 'rollup-plugin-terser';
import typescript from 'rollup-plugin-typescript2';
import pkg from './package.json';
import terserOptions from './terser.json';

export default [
  {
    input: 'src/index.ts',
    output: [
      {
        exports: 'named',
        file: 'dist/index.js',
        format: 'umd',
        name: 'BloomreachSpaSdk',
        sourcemap: true,
        sourcemapExcludeSources: true,
      },
      {
        file: 'dist/index.mjs',
        format: 'esm',
      },
    ],
    external: [
      ...Object.keys(pkg.dependencies || {}),
      ...Object.keys(pkg.peerDependencies || {}),
    ],
    plugins: [
      typescript({
        cacheRoot: './node_modules/.cache/rpt2',
        useTsconfigDeclarationDir: true,
      }),
      babel({ extensions: ['.ts'] }),
      terser(terserOptions)
    ],
  },

  {
    input: 'src/index.ts',
    output: [{
      file: 'dist/index.es6.js',
      format: 'esm',
    }],
    external: [
      ...Object.keys(pkg.dependencies || {}),
      ...Object.keys(pkg.peerDependencies || {}),
    ],
    plugins: [
      typescript({
        cacheRoot: './node_modules/.cache/rpt2',
        tsconfigOverride: {
          compilerOptions: {
            declaration: false,
          }
        },
      }),
      terser(terserOptions)
    ],
  },

  {
    input: 'src/index.ts',
    output: [{
      file: 'dist/index.d.ts',
      format: 'es',
    }],
    external: [
      ...Object.keys(pkg.dependencies || {}),
      ...Object.keys(pkg.peerDependencies || {}),
    ],
    plugins: [
      dts(),
    ],
  },
];
