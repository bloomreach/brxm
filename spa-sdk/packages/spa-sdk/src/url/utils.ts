/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

interface Url {
  hash: string;
  origin: string;
  pathname: string;
  search: string;
  searchParams: URLSearchParams;
  path: string;
}

export function appendSearchParams(url: string, params: URLSearchParams) {
  const { hash, origin, pathname, searchParams } = parseUrl(url);

  return buildUrl({ hash, origin, pathname, searchParams: mergeSearchParams(searchParams, params) });
}

export function buildUrl(url: Partial<Url>) {
  const searchParams = url.searchParams?.toString() ?? '';
  const search = url.search ?? `${searchParams && `?${searchParams}`}`;
  const path = url.path ?? `${url.pathname ?? ''}${search}${url.hash ?? ''}`;

  return `${url.origin ?? ''}${path}`;
}

export function extractSearchParams(url: string, params: string[]) {
  const { hash, origin, pathname, searchParams } = parseUrl(url);
  const extracted = new URLSearchParams();

  params.forEach((param) => {
    if (searchParams.has(param)) {
      extracted.set(param, searchParams.get(param)!);
      searchParams.delete(param);
    }
  });

  return {
    searchParams: extracted,
    url: buildUrl({ hash, origin, pathname, searchParams }),
  };
}

export function isAbsoluteUrl(url: string): boolean {
  const { origin, pathname } = parseUrl(url);

  return !!origin || pathname.startsWith('/');
}

function isMatchedOrigin(origin: string, baseOrigin: string) {
  const [schema, host = ''] = origin.split('//', 2);
  const [baseSchema, baseHost = ''] = baseOrigin.split('//', 2);

  return !baseOrigin
    || !origin
    || (!schema || !baseSchema || schema === baseSchema) && baseHost === host;
}

function isMatchedPathname(pathname: string, basePathname: string) {
  return !basePathname || pathname.startsWith(basePathname);
}

function isMatchedQuery(search: URLSearchParams, baseSearch: URLSearchParams) {
  let match = true;
  baseSearch.forEach((value, key) => {
    match = match && (!value && search.has(key) || search.getAll(key).includes(value));
  });

  return match;
}

export function isMatched(link: string, base = '') {
  const linkUrl = parseUrl(link);
  const baseUrl = parseUrl(base);

  return isMatchedOrigin(linkUrl.origin, baseUrl.origin)
    && isMatchedPathname(linkUrl.pathname, baseUrl.pathname)
    && isMatchedQuery(linkUrl.searchParams, baseUrl.searchParams);
}

export function mergeSearchParams(params: URLSearchParams, ...rest: URLSearchParams[]) {
  const result = new URLSearchParams(params);
  rest.forEach(params => params.forEach((value, key) => result.set(key, value)));

  return result;
}

export function parseUrl(url: string): Url {
  // URL constructor requires either a valid URL or a base URL.
  // Since this function returns a pathname, we can safely pass a fake host to be able to resolve relative URLs.
  const parsedUrl = url ? new URL(url, 'http://example.com') : {} as URL;
  const { hash = '', search = '', searchParams = new URLSearchParams() } = parsedUrl;

  // For links like `//example.com?query#hash` pathname will be `/` so we need to strip query and hash parameters first.
  let origin = url.substring(0, url.length - search.length - hash.length);
  let { pathname = '' } = parsedUrl;
  if (!origin.endsWith(pathname)) {
    pathname = pathname.substring(1);
  }

  origin = origin.substring(0, origin.length - pathname.length);

  return { hash, origin, pathname, search, searchParams, path: `${pathname}${search}${hash}` };
}

export function resolveUrl(url: string, base: string) {
  const baseUrl = parseUrl(base);
  const sourceUrl = parseUrl(url);
  const pathname = sourceUrl.pathname.startsWith('/')
    ? sourceUrl.pathname
    : `${baseUrl.pathname}${baseUrl.pathname.endsWith('/') || !sourceUrl.pathname ? '' : '/'}${sourceUrl.pathname}`;

  return buildUrl({
    pathname,
    hash: sourceUrl.hash || baseUrl.hash,
    origin: sourceUrl.origin || baseUrl.origin,
    searchParams: mergeSearchParams(baseUrl.searchParams, sourceUrl.searchParams),
  });
}
