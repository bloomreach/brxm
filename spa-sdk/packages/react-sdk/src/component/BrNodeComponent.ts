/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { Component } from '@bloomreach/spa-sdk';
import { BrProps } from './BrProps';
import { BrMappingContext } from './BrMappingContext';

export class BrNodeComponent<T extends Component> extends React.Component<BrProps<T>> {
  static contextType = BrMappingContext;
  context!: React.ContextType<typeof BrMappingContext>;

  protected getMapping(): string | undefined {
    return this.props.component.getName();
  }

  protected fallback() {
    return this.props.children;
  }

  render() {
    const mapping = this.getMapping();
    const component = mapping && this.context[mapping];
    if (!component) {
      return this.fallback();
    }

    return React.createElement(component, this.props);
  }
}
