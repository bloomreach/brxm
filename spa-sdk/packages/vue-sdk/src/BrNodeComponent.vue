<!--
  Copyright 2020 Hippo B.V. (http://www.onehippo.com)

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
  -->

<template>
  <br-meta :meta="component.getMeta()">
    <slot>
      <br-node-container-item v-if="isContainerItem(component)" />

      <br-node-container v-else-if="isContainer(component)">
        <br-node-component v-for="(component, key) in component.getChildren()" :key="key" :component="component" />
      </br-node-container>

      <component
        v-else-if="component.getName() in mapping"
        :is="mapping[component.getName()]"
        :component="component"
        :page="page"
      />

      <br-node-component v-else v-for="(component, key) in component.getChildren()" :key="key" :component="component" />
    </slot>
  </br-meta>
</template>

<script lang="ts">
import { Component as SpaComponent, Page, isContainerItem, isContainer } from '@bloomreach/spa-sdk';
import { Component, Inject, Prop, Provide, Vue } from 'vue-property-decorator';
import BrMeta from './BrMeta.vue';
import BrNodeContainerItem from './BrNodeContainerItem.vue';
import BrNodeContainer from './BrNodeContainer.vue';

@Component({
  components: {
    BrMeta,
    BrNodeContainerItem,
    BrNodeContainer,
  },
  computed: {
    mapping(this: BrNodeComponent) {
      return this.mapping$();
    },
    page(this: BrNodeComponent) {
      return this.page$();
    },
  },
  methods: {
    isContainerItem,
    isContainer,
  },
  name: 'br-node-component',
})
export default class BrNodeComponent extends Vue {
  @Prop() component!: SpaComponent;

  @Inject() private mapping$!: () => Record<string, Vue.Component>;

  @Inject() private page$!: () => Page;

  @Provide() // ProvideReactive doesn't work with recursive components
  private component$() {
    return this.component;
  }
}
</script>
