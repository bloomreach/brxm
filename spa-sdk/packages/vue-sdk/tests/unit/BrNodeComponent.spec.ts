/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { mocked } from 'ts-jest/utils';
import { Wrapper, shallowMount } from '@vue/test-utils';
import { Component, Prop, Provide, Vue } from 'vue-property-decorator';
import { Component as BrComponent, MetaCollection, Page, isContainerItem, isContainer } from '@bloomreach/spa-sdk';
import BrNodeComponent from '@/BrNodeComponent.vue';

jest.mock('@bloomreach/spa-sdk');

@Component({ template: '<div />' })
class BrPage extends Vue {
  @Prop() mapping!: Record<string, Vue.Component>;

  @Provide() mapping$() {
    return this.mapping;
  }

  @Prop() page!: Page;

  @Provide() page$() {
    return this.page;
  }
}

@Component
class SomeComponent extends Vue {
  @Prop() component!: BrComponent;

  @Prop() page!: Page;
}

describe('BrNodeComponent', () => {
  let component: jest.Mocked<BrComponent>;
  let mapping: Record<string, Vue.Component>;
  let page: jest.Mocked<Page>;
  let provide: Function;
  let shallowMountComponent: () => Wrapper<Vue>;

  beforeEach(() => {
    component = ({
      getChildren: jest.fn(() => []),
      getMeta: jest.fn(),
      getName: jest.fn(),
    } as unknown) as typeof component;
    mapping = { something: SomeComponent };
    page = ({} as unknown) as typeof page;

    const wrapper = shallowMount(BrPage, { propsData: { mapping, page } });
    provide = (wrapper.vm.$options.provide as Function).bind(wrapper.vm);

    shallowMountComponent = () => shallowMount(BrNodeComponent, { provide, propsData: { component } });
  });

  describe('render', () => {
    it('should surround the component with meta-data', () => {
      component.getMeta.mockReturnValue((['a', 'b'] as unknown) as MetaCollection);
      const wrapper = shallowMountComponent();
      const metaComponent = wrapper.findComponent({ name: 'br-meta' });

      expect(metaComponent.props()).toEqual({ meta: ['a', 'b'] });
    });

    it('should render a container item', async () => {
      mocked(isContainerItem).mockReturnValueOnce(true);

      const wrapper = shallowMountComponent();
      await wrapper.vm.$nextTick();

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render a container', async () => {
      const child1 = ({ getName: () => 'child1' } as unknown) as BrComponent;
      const child2 = ({ getName: () => 'child2' } as unknown) as BrComponent;
      component.getChildren.mockReturnValue([child1, child2]);
      mocked(isContainer).mockReturnValueOnce(true);

      const wrapper = shallowMountComponent();
      await wrapper.vm.$nextTick();

      const children = wrapper.findAllComponents(BrNodeComponent);

      expect(wrapper.html()).toMatchSnapshot();
      expect(children.at(1).props().component.getName()).toBe('child1');
      expect(children.at(2).props().component.getName()).toBe('child2');
    });

    it('should render a mapped component', async () => {
      component.getName.mockReturnValue('something');

      const wrapper = shallowMountComponent();
      await wrapper.vm.$nextTick();

      const props = wrapper.findComponent(SomeComponent).props();

      expect(wrapper.html()).toMatchSnapshot();
      expect(props.component).toBe(component);
      expect(props.page).toEqual(page);
    });

    it('should render children', async () => {
      const child1 = ({ getName: () => 'child1' } as unknown) as BrComponent;
      const child2 = ({ getName: () => 'child2' } as unknown) as BrComponent;
      component.getChildren.mockReturnValue([child1, child2]);

      const wrapper = shallowMountComponent();
      await wrapper.vm.$nextTick();

      const children = wrapper.findAllComponents(BrNodeComponent);

      expect(wrapper.html()).toMatchSnapshot();
      expect(children.at(1).props().component.getName()).toBe('child1');
      expect(children.at(2).props().component.getName()).toBe('child2');
    });
  });
});
