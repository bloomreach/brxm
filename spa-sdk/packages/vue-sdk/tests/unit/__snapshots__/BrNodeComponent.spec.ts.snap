// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrNodeComponent render should render a container 1`] = `
<br-meta-stub>
  <br-node-container-stub>
    <br-node-component-stub component="[object Object]"></br-node-component-stub>
    <br-node-component-stub component="[object Object]"></br-node-component-stub>
  </br-node-container-stub>
</br-meta-stub>
`;

exports[`BrNodeComponent render should render a container item 1`] = `
<br-meta-stub>
  <br-node-container-item-stub></br-node-container-item-stub>
</br-meta-stub>
`;

exports[`BrNodeComponent render should render a mapped component 1`] = `
<br-meta-stub>
  <somecomponent-stub component="[object Object]" page="[object Object]"></somecomponent-stub>
</br-meta-stub>
`;

exports[`BrNodeComponent render should render children 1`] = `
<br-meta-stub>
  <br-node-component-stub component="[object Object]"></br-node-component-stub>
  <br-node-component-stub component="[object Object]"></br-node-component-stub>
</br-meta-stub>
`;
